### ISSUE TITLE
conda execute a Jupyter notebook

### ISSUE THREAD

**Issue by pelson** [pelson/conda-execute#3]

**Comment by pelson** [comment]
Original thread came from https://twitter.com/mrocklin/status/655100657016242176


> @Carreau [mentioned](https://twitter.com/Mbussonn/status/655228658215350273) the ``nbconvert --execute`` option.

It seems nbconvert is already in this space, the only way I would want ``conda-execute`` to add anything is if we can make use of the fact that we can define a full environment. It seems that ```conda execute my_analysis.ipynb``` could do a number of things:

 * launch the Jupyter notebook with a suitable kernel without running any of the actual code
 * run the notebook and convert to html (as nbconvert is currently doing)
 * run the notebook and save on top of the existing notebook

Rather than bake any of this in, ``conda-execute`` would essentially call out to the appropriate executable:

```
$ cat my_analysis.ipynb
{
  "metadata" : {
    "conda-execute": {
        "env": ['python >2', 'numpy==1.10', 'nbconvert'],
        "run_with": ["nbconvert",  "--execute"]
    },
  },
  "nbformat": 4,
  "nbformat_minor": 0,
  "cells" : [
  ...
   ]
}
```

Of course, to do this, conda-execute *does* need to read the metadata from the notebook format.

**Comment by takluyver** [comment]
I think that it would be valuable to have conda setting up environments based on notebook metadata. One way to do this might be a custom kernelspec manager which could look at metadata when a notebook is opened and construct a suitable environment at that time. I think we've talked to @ahmadia about a similar idea w/r/t hashdist.

**Comment by ahmadia** [comment]
Hi folks.  We've been trying to figure out how to do this from the conda side as well.  As @takluyver mentioned, one option is to attach the conda environment directly to the notebook metadata.  See the discussion here: https://github.com/conda/conda-env/pull/111 and the feature here: http://conda.pydata.org/docs/commands/env/conda-env-attach.html

From there, you need a corresponding launch action.  One option is conda launch from @ijstokes - https://github.com/conda/conda-launch

cc: @chdoig @damianavila @pwang @malev

I'm happy to discuss here a bit more about what the various options are and what we like/dislike about each.

**Comment by ahmadia** [comment]
Of course, anything we do for conda should be similarly applicable to HashDist.  Thanks for the mention @takluyver.

**Comment by Carreau** [comment]
For `nbconvert --execute` you need to replace the [execute preprocessor](https://github.com/jupyter/nbconvert/blob/master/nbconvert/preprocessors/execute.py#L33) by one that start kernels in an env. Kernelspec already can have an "env" metadata so overwriting PATH, should be sufficient. 

Running the notebook app in the env seem unnecessary as IIUC, you want to run the notebook kernel in the env, 

> run the notebook and convert to html (as nbconvert is currently doing)
> run the notebook and save on top of the existing notebook

This will already be taken care of if you use `--to ipynb` or `--to html` with nbconvert. 
I guess you might want to use nbconvert as a library instead of shelling out, we can figure that out.

**Comment by ahmadia** [comment]
> Running the notebook app in the env seem unnecessary as IIUC, you want to run the notebook kernel in the env,

This made sense when the kernel spec was changing rapidly and you wanted the app version to match the kernel version.  I agree that this is not a hard requirement for using environments.
